### PR TITLE
feat(groovy)!: switch to more complete parser

### PR DESCRIPTION
**Pull request by murtaza64**
I've been actively building a new parser for groovy because I was dissatisfied with the included one. The [README of my project](https://github.com/murtaza64/tree-sitter-groovy/blob/main/README.md) includes a screenshot showing some of the superficial improvements, but this new parser much more accurately and richly parses groovy code based on the groovy language spec, and supports features including annotations, closures, the various string types in groovy and more. It also includes a pretty extensive test suite to aid in future improvement.

I've been daily driving it for a few months now, and I work on Jenkins pipelines a lot and have been pretty happy with it.

Willing to change or add stuff if other people find issues with it. Also willing to go into more detail about the improvements, but wanted to put this PR up to start the discussion.

I've been [discussing](https://github.com/Decodetalkers/tree-sitter-groovy/issues/5) with @Decodetalkers about their interest in adopting my parser--they wanted me to submit a PR to their repo, but since this was a full rewrite of the parser, I'm waiting for confirmation that they're ok with replacing the upstream parser here.

Thanks @WillLillis for your help here too!

### CODE REVIEW

**Comment by clason** [comment]
No injections, not even comments?

And please check CONTRIBUTING.md for the list of allowed capture names; we enforce these in order to have consistency across languages (necessary for colorscheme authors).

**Comment by murtaza64** [comment]
> No injections, not even comments?

What's the benefit of having injections for `comment`? I implemented some parsing of comment content to find interpolations and escape sequences--wouldn't an injection for comments mess that up?

**Comment by clason** [comment]
> I'm seeing the CI failures... Not sure why `@parameter` works but `@variable.parameter` doesn't seem to highlight parameters in red in my theme, even though for other languages like python, parameters do get highlighted and their queries are set up to be `@variable.parameter`.

Use `:Inspect` to see the applied highlights with their order.

**Comment by murtaza64** [comment]
> Use :Inspect to see the applied highlights with their order.

It looks like the python queries on my machine hadn't been updated to use `@variable.parameter` instead of `@parameter`, and neither had my color theme. Fixed now.

**Comment by Decodetalkers** [comment]
+1

File: queries/groovy/highlights.scm [hunk 1]
```diff
@@ -1,70 +1,261 @@
...
+  "private"
+  "protected"
+  "public"
+  "static"
+  "synchronized"
+] @type.qualifier
 
-((identifier) @boolean
-  (#any-of? @boolean "true" "false" "True" "False"))
+(comment) @comment
```
Comment (lucario387) [inline queries/groovy/highlights.scm:58]: Are there specs for documentation comments? If yes, then having a capture for `@comment.documentation` will be nice

File: queries/groovy/highlights.scm [hunk 2]
```diff
@@ -1,70 +1,261 @@
...
-  (#any-of? @keyword "static" "class" "def" "import" "package" "assert" "extends" "implements" "instanceof" "interface" "new"))
+"(" @punctuation.bracket
 
-((identifier) @type.qualifier
-  (#any-of? @type.qualifier "abstract" "protected" "private" "public"))
+")" @punctuation.bracket
 
-((identifier) @keyword.exception
-  (#any-of? @keyword.exception "throw" "finally" "try" "catch"))
+"[" @punctuation.bracket
```
Comment (lucario387) [inline queries/groovy/highlights.scm:75]: Group all of these `@punctuation.bracket` into one

e.g.
```
[
  ...
]
```

File: queries/groovy/highlights.scm (hunk 1 above)
Comment (lucario387) [reply queries/groovy/highlights.scm:58]: Ah nice ! :tada: I didn't see `groovy_doc` below

**Review by lucario387 — COMMENTED** [review]
What about `folds` :thinking:

File: queries/groovy/highlights.scm [hunk 3]
```diff
@@ -1,70 +1,261 @@
...
+  function: (identifier) @function)
+
+(annotation) @function.macro
+
+(annotation
+  (identifier) @function.macro)
+
+"@interface" @function.macro
+
+"pipeline" @keyword
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:244]: Highlights specific to Jenkinsfile should be in a Jenkinsfile parser that inherits this one, or the user's files.

File: queries/groovy/highlights.scm [hunk 4]
```diff
@@ -1,70 +1,261 @@
...
+  "instanceof"
+  "package"
+  "pipeline"
+  "return"
+  "switch"
+  "try"
+  "while"
+  (break)
+  (continue)
+] @keyword
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:27]: Some of these can be more specific (e.g. `if` -> `@keyword.conditional`).

File: queries/groovy/highlights.scm [hunk 5]
```diff
@@ -1,70 +1,261 @@
...
 
-(string
-  (identifier) @variable)
+[
+  "true"
+  "false"
+] @boolean
 
-(escape_sequence) @string.escape
+(null) @constant
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:34]:
```suggestion
(null) @constant.builtin
```

File: queries/groovy/highlights.scm (hunk 1 above)
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:58]:
```suggestion
(comment) @comment @spell
```

File: queries/groovy/highlights.scm [hunk 6]
```diff
@@ -1,70 +1,261 @@
...
+  "synchronized"
+] @type.qualifier
 
-((identifier) @boolean
-  (#any-of? @boolean "true" "false" "True" "False"))
+(comment) @comment
 
-((identifier) @constant
-  (#lua-match? @constant "^[A-Z][A-Z%d_]*$"))
+(shebang) @comment
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:60]:
```suggestion
(shebang) @keyword.directive
```

File: queries/groovy/highlights.scm [hunk 7]
```diff
@@ -1,70 +1,261 @@
...
-((identifier) @type
-  (#any-of? @type "String" "Map" "Object" "Boolean" "Integer" "List"))
+(string
+  (escape_sequence) @operator)
 
-((identifier) @function.builtin
-  (#any-of? @function.builtin "void" "id" "version" "apply" "implementation" "testImplementation" "androidTestImplementation" "debugImplementation"))
+(string
+  (interpolation
+    "$" @operator))
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:69]:
```suggestion
    "$" @punctuation.special))
```

File: queries/groovy/highlights.scm [hunk 8]
```diff
@@ -1,70 +1,261 @@
...
+(shebang) @comment
 
-((identifier) @constant.builtin
-  (#eq? @constant.builtin "null"))
+(string) @string
 
-((identifier) @type
-  (#any-of? @type "String" "Map" "Object" "Boolean" "Integer" "List"))
+(string
+  (escape_sequence) @operator)
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:65]:
```suggestion
  (escape_sequence) @string.escape)
```

File: queries/groovy/highlights.scm [hunk 9]
```diff
@@ -1,70 +1,261 @@
...
+(annotation) @function.macro
+
+(annotation
+  (identifier) @function.macro)
+
+"@interface" @function.macro
+
+"pipeline" @keyword
+
+(groovy_doc) @comment.documentation
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:246]:
```suggestion
(groovy_doc) @comment.documentation @spell
```

File: queries/groovy/highlights.scm [hunk 10]
```diff
@@ -1,70 +1,261 @@
...
+
+(groovy_doc
+  [
+    (groovy_doc_param)
+    (groovy_doc_throws)
+    (groovy_doc_tag)
+  ] @string.special)
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:253]:
```suggestion
  ] @string.special @nospell)
```

File: queries/groovy/highlights.scm [hunk 11]
```diff
@@ -1,70 +1,261 @@
...
+
+(groovy_doc
+  (groovy_doc_param
+    (identifier) @variable.parameter))
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:257]:
```suggestion
    (identifier) @variable.parameter @nospell))
```

File: queries/groovy/highlights.scm [hunk 12]
```diff
@@ -1,70 +1,261 @@
...
+
+(groovy_doc
+  (groovy_doc_throws
+    (identifier) @type))
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:261]:
```suggestion
    (identifier) @type @nospell))
```

File: queries/groovy/highlights.scm (hunk 2 above)
Comment (ObserverOfTime) [reply queries/groovy/highlights.scm:75]: Same for `@punctuation.delimiter`.

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

**Comment by murtaza64** [comment]
> What about folds 🤔
I suppose I could work on it, I didn't because I never use them.

Thanks for the detailed review, will address soon

**Comment by murtaza64** [comment]
@ObserverOfTime Tried to address your comments as best I could :)

@lucario387 added some basic folds

**Comment by murtaza64** [comment]
> Highlights specific to Jenkinsfile should be in a Jenkinsfile parser that inherits this one, or the user's files.

do you have an example of a parser that inherits another?

**Comment by clason** [comment]
The whole CPP/GSL/etc. family.

**Comment by ObserverOfTime** [comment]
[HCL/Terraform](https://github.com/MichaHoffmann/tree-sitter-hcl) is the best example here since there's no actual difference in the grammars, just the queries.
(You might as well make one for Gradle too while you're at it. We can help with the queries.)

File: queries/groovy/folds.scm [hunk 13]
```diff
@@ -0,0 +1,4 @@
+[
+  (closure)
+  (argument_list)
+] @fold
```
Comment (ObserverOfTime) [inline queries/groovy/folds.scm:4]: What about lists, maps, and blocks?

File: queries/groovy/highlights.scm [hunk 14]
```diff
@@ -1,64 +1,83 @@
...
+  "assert"
+  "case"
+  "class"
+  "default"
+  "extends"
+  "in"
+  "instanceof"
+  "package"
+  "switch"
+] @keyword
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:14]: "in", "!in", "as" should be `@keyword.operator`.
"switch", "default", "case" should be in `@keyword.conditional`.

File: queries/groovy/highlights.scm [hunk 15]
```diff
@@ -68,3 +87,179 @@
...
+] @operator
+
+(string
+  "/" @string)
+
+(ternary_op
+  ([
+    "?"
+    ":"
+  ]) @operator)
```
Comment (ObserverOfTime) [inline queries/groovy/highlights.scm:154]:
```suggestion
  ]) @conditional.ternary)
```

File: queries/groovy/highlights.scm (hunk 8 above)
Comment (ObserverOfTime) [reply queries/groovy/highlights.scm:76]: You forgot this one.

File: queries/groovy/highlights.scm (hunk 7 above)
Comment (ObserverOfTime) [reply queries/groovy/highlights.scm:80]: And this one.

File: queries/groovy/folds.scm (hunk 13 above)
Comment (murtaza64) [reply queries/groovy/folds.scm:4]: All blocks are closures in my grammar. will throw in lists and maps

**Review by ObserverOfTime — APPROVED** [review]

**Comment by murtaza64** [comment]
Yay! thanks all for the help :)

**Comment by murtaza64** [comment]
Will I need to make PRs for updates/improvements/fixes to the parser?

**Comment by amaanq** [comment]
Only if there's anything breaking, then *please do*, otherwise you can just leave it be (though any new features that would result in some additions to the queries added would be great to port here)

**Pull request merged**